jwt/keyset_test.go

```diff
@@ -1 +1,696 @@
+			args: args{
+				jwksURL:   tp.Addr() + wellKnownJWKS,
+				jwksCAPEM: tp.CACert(),
+			},
+		},
+		{
+			name: "empty JWKS URL",
+			args: args{
+				jwksURL: "",
+			},
+			wantErr: true,
+		},
+		{
+			name: "malformed JWKS CA PEM",
+			args: args{
+				jwksURL:   tp.Addr() + wellKnownJWKS,
+				jwksCAPEM: "-----BEGIN CERTIFICATE-----",
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewJSONWebKeySet(context.Background(), tt.args.jwksURL, tt.args.jwksCAPEM)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestNewOIDCDiscoveryKeySet(t *testing.T) {
+	tp := oidc.StartTestProvider(t)
+
+	type args struct {
+		issuer             string
+		issuerCAPEM        string
+		modifyTestProvider func()
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "valid issuer and CA PEM",
+			args: args{
+				issuer:      tp.Addr(),
+				issuerCAPEM: tp.CACert(),
+			},
+		},
+		{
+			name: "empty issuer URL",
+			args: args{
+				issuer: "",
+			},
+			wantErr: true,
+		},
+		{
+			name: "invalid issuer URL",
+			args: args{
+				issuer: "https:example.com/",
+			},
+			wantErr: true,
+		},
+		{
+			name: "malformed issuer CA PEM",
+			args: args{
+				issuer:      tp.Addr(),
+				issuerCAPEM: "-----BEGIN CERTIFICATE-----",
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewOIDCDiscoveryKeySet(context.Background(), tt.args.issuer, tt.args.issuerCAPEM)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestNewStaticKeySet(t *testing.T) {
+	type args struct {
+		publicKeys func() []crypto.PublicKey
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "valid public keys",
+			args: args{
+				publicKeys: func() []crypto.PublicKey {
+					priv, err := rsa.GenerateKey(rand.Reader, 2048)
+					require.NoError(t, err)
+					return []crypto.PublicKey{priv.Public()}
+				},
+			},
+		},
+		{
+			name: "empty public keys",
+			args: args{
+				publicKeys: func() []crypto.PublicKey {
+					return nil
+				},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewStaticKeySet(tt.args.publicKeys())
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+		})
+	}
+}
+
+func testJWTClaims(t *testing.T) map[string]interface{} {
+	t.Helper()
+	return map[string]interface{}{
+		"iss": "https://example.com/",
+		"sub": "alice@example.com",
+		"aud": []interface{}{"www.example.com"},
+		"exp": float64(1611699944),
+		"nbf": float64(1611699344),
+		"iat": float64(1611699344),
+		"jti": "abc123",
+	}
+}
+
+func testSignJWT(t *testing.T, key crypto.PrivateKey, alg Alg, claims interface{}, keyID []byte) string {
```

Comment:
Should we move this to testing.go and make it part of the packages test API?  That would allow you to use it in `docs_test.go` for test tokens.

Reply:
I didn't realize that there was already a [TestSignJWT()](https://github.com/hashicorp/cap/blob/main/oidc/testing.go#L38) in testing.go. I chose to adopt using that in the jwt tests instead. I ended up changing the `TestSignJWT()` signature to take a string for an alg. I did this so that both `jwt.Alg` and `oidc.Alg` can be used in tests. Let me know your thoughts!

Reply:
I like these changes.